File: docs/.vitepress/theme/samples.ts
/**
 * The pairs of documents the demos compare.
 *
 * Written here rather than in the pages because a Markdown page cannot carry a
 * multi-line string into a component without turning into an escaping puzzle,
 * and because every page showing the same pair is what makes the options
 * comparable: a reader turning wrapping on should see the wrapping change and
 * nothing else.
 *
 * Each pair is longer than it has to be to show one change, and deliberately:
 * the playground is somewhere to scroll a comparison, fold it, search it and
 * step between changes, and a pair that fits on one screen answers none of
 * those. So the changes are spread from the top of each document to the bottom
 * rather than gathered where they would all be visible at once.
 */

export interface Sample {
  before: string;
  after: string;
  beforeLabel: string;
  afterLabel: string;
  /** What the pair is written in, for `language`. Left out where it is prose. */
  language?: string;
}

const CODE_BEFORE = `const TAX_RATE = 0.08;

export function subtotal(items) {
  let total = 0;

  for (const item of items) {
    total += item.price * item.quantity;
  }

  return total;
}

export function tax(items) {
  return subtotal(items) * TAX_RATE;
}

export function total(items) {
  return subtotal(items) + tax(items);
}

// Rounding is the caller's problem for now.
export function format(amount) {
  return '$' + amount.toFixed(2);
}

export function addItem(cart, product, quantity = 1) {
  const line = cart.lines.find((entry) => entry.sku === product.sku);

  if (line) {
    line.quantity += quantity;

    return cart;
  }

  cart.lines.push({
    sku: product.sku,
    name: product.name,
    price: product.price,
    quantity
  });

  return cart;
}

export function removeItem(cart, sku) {
  cart.lines = cart.lines.filter((line) => line.sku !== sku);

  return cart;
}

export function setQuantity(cart, sku, quantity) {
  const line = cart.lines.find((entry) => entry.sku === sku);

  if (!line) {
    return cart;
  }

  if (quantity < 1) {
    return removeItem(cart, sku);
  }

  line.quantity = quantity;

  return cart;
}

export function itemCount(cart) {
  return cart.lines.reduce((sum, line) => sum + line.quantity, 0);
}

export function isEmpty(cart) {
  return cart.lines.length === 0;
}

export function summarise(cart) {
  return {
    lines: cart.lines.length,
    items: itemCount(cart),
    total: format(total(cart.lines))
  };
}
`;

const CODE_AFTER = `const DECIMALS = { USD: 2, EUR: 2, JPY: 0, KRW: 0 };
const TAX_RATES = { US: 0.08, DE: 0.19, JP: 0.1, KR: 0.1 };

export function subtotal(items, currency = 'USD') {
  let total = 0;

  for (const item of items) {
    total += item.price * item.quantity;
  }

  return round(total, currency);
}

function round(amount, currency) {
  return Number(amount.toFixed(DECIMALS[currency] ?? 2));
}

export function tax(items, region = 'US', currency = 'USD') {
  return round(subtotal(items, currency) * (TAX_RATES[region] ?? 0), currency);
}

export function total(items, region = 'US', currency = 'USD') {
  return round(subtotal(items, currency) + tax(items, region, currency), currency);
}

export function format(amount, currency = 'USD') {
  return new Intl.NumberFormat('en-US', { style: 'currency', currency }).format(amount);
}

export function addItem(cart, product, quantity = 1) {
  const line = cart.lines.find((entry) => entry.sku === product.sku);

  if (line) {
    line.quantity += quantity;

    return cart;
  }

  cart.lines.push({
    sku: product.sku,
    name: product.name,
    price: product.price,
    currency: product.currency ?? cart.currency,
    quantity
  });

  return cart;
}

export function removeItem(cart, sku) {
  cart.lines = cart.lines.filter((line) => line.sku !== sku);

  return cart;
}

export function setQuantity(cart, sku, quantity) {
  const line = cart.lines.find((entry) => entry.sku === sku);

  if (!line) {
    return cart;
  }

  if (quantity < 1) {
    return removeItem(cart, sku);
  }

  line.quantity = quantity;

  return cart;
}

export function applyDiscount(cart, code) {
  const rule = cart.discounts?.[code];

  if (!rule) {
    return cart;
  }

  cart.discount = { code, off: rule.off };

  return cart;
}

export function itemCount(cart) {
  return cart.lines.reduce((sum, line) => sum + line.quantity, 0);
}

export function summarise(cart) {
  return {
    lines: cart.lines.length,
    items: itemCount(cart),
    currency: cart.currency,
    total: format(total(cart.lines, cart.region, cart.currency), cart.currency)
  };
}
`;

const PROSE_BEFORE = `The Release Notes

We have shipped a new version of the editor. It fixes a number of bugs and adds
one feature that people have asked for.

What is new

The new feature is a keyboard shortcut for saving. Press it and the document is
written to disk. There is nothing to set up first.

We made the sidebar narrower, which gives the text more room on a small screen.

The status bar counts words as well as lines.

What we fixed

Opening a file with no final newline no longer adds one.

Search stops at the end of the document instead of wrapping around without
saying so.

The window remembers its size between sessions.

Pasting a wide table no longer takes several seconds.

A document left open overnight no longer loses its undo history.

What is still wrong

Printing comes out wrong on a page that has been zoomed. We know why, and it is
the next thing on the list.

The spell checker only knows one dictionary at a time.

Upgrading

Your settings are carried over. Nothing has to be installed again and no file
format has changed.

Let us know what you think.
`;

const PROSE_AFTER = `Release notes for 2.4

We have shipped a new version of the editor. It fixes eleven bugs and adds two
features that people have asked for.

What is new

The first is a keyboard shortcut for saving. Press it and the document is
written to disk. There is nothing to set up first.

The second is an autosave that runs every thirty seconds and stays out of the
way. It is on for a new installation and can be turned off under Preferences.

We made the sidebar narrower, which gives the text more room on a small screen.
It can be dragged back to where it was.

The status bar counts words, lines and characters, and says which of the three
is selected.

What we fixed

Opening a file with no final newline no longer adds one.

Search stops at the end of the document and says so, rather than wrapping
around in silence.

The window remembers its size and its position between sessions.

Pasting a wide table is about forty times faster than it was.

A document left open overnight no longer loses its undo history.

What is still wrong

Printing comes out wrong on a page that has been zoomed. We know why, and it is
the next thing on the list.

Upgrading

Your settings are carried over. Nothing has to be installed again and no file
format has changed.

Thank you to everybody who wrote in about the search bar. Let us know what you
make of this one.
`;

const CONFIG_BEFORE = `name: deploy

on:
  push:
    branches: [main]

jobs:
  check:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 20
          cache: npm
      - run: npm ci
      - run: npm run lint
      - run: npm run typecheck

  build:
    needs: check
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 20
          cache: npm
      - run: npm ci
      - run: npm test
      - run: npm run build
      - uses: actions/upload-artifact@v4
        with:
          name: site
          path: dist

  deploy:
    needs: build
    runs-on: ubuntu-latest
    steps:
      - uses: actions/download-artifact@v4
        with:
          name: site
          path: dist
      - run: ./scripts/upload.sh dist
`;

const CONFIG_AFTER = `name: deploy

on:
  push:
    branches: [main, release/*]
  workflow_dispatch:

concurrency:
  group: deploy
  cancel-in-progress: true

jobs:
  check:
    runs-on: ubuntu-24.04
    timeout-minutes: 10
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 22
          cache: npm
      - run: npm ci
      - run: npm run lint
      - run: npm run typecheck
      - run: npm run format -- --check

  build:
    needs: check
    runs-on: ubuntu-24.04
    timeout-minutes: 15
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 22
          cache: npm
      - run: npm ci
      - run: npm test -- --coverage
      - run: npm run build
      - uses: actions/upload-artifact@v4
        with:
          name: site
          path: dist
          retention-days: 7

  deploy:
    needs: build
    runs-on: ubuntu-24.04
    environment: production
    steps:
      - uses: actions/download-artifact@v4
        with:
          name: site
          path: dist
      - run: ./scripts/upload.sh dist
      - run: ./scripts/purge-cache.sh
`;

const KOREAN_BEFORE = `공원 이용 안내

개장 시간

여름철 개장 시간은 오전 6시부터 오후 10시까지입니다.
겨울철에는 오전 7시부터 오후 8시까지 문을 엽니다.
정기 휴원일은 매월 첫째 주 월요일입니다.

반려동물

반려동물은 목줄을 채우면 들어올 수 있습니다.
배변 봉투는 각 출입구에 놓아 두었습니다.
잔디 광장에는 반려동물을 데리고 들어갈 수 없습니다.

자전거와 주차

자전거는 동문 주차장에 세워 주세요.
주차장은 두 시간까지 무료이고, 그 뒤로는 30분에 500원입니다.
대형 버스는 미리 연락해 자리를 확인해 주세요.

시설 이용

야외 취사와 음주는 할 수 없습니다.
분수 광장은 매시 정각에 15분 동안 가동합니다.
화장실은 정문, 동문, 호수 쉼터 세 곳에 있습니다.

문의

공원 관리 사무소는 정문 옆 1층에 있습니다.
평일 오전 9시부터 오후 6시까지 문을 엽니다.
`;

const KOREAN_AFTER = `공원 이용 안내

개장 시간

여름철 개장 시간은 오전 5시부터 오후 11시까지입니다.
겨울철에는 오전 7시부터 오후 8시까지 문을 엽니다.
정기 휴원일은 매월 첫째 주와 셋째 주 월요일입니다.

반려동물

반려동물은 목줄을 채우면 들어올 수 있습니다.
배변 봉투는 각 출입구와 쉼터에 놓아 두었습니다.
잔디 광장에는 반려동물을 데리고 들어갈 수 없습니다.
맹견은 입마개를 씌워 주세요.

자전거와 주차

자전거는 동문과 서문 주차장에 세워 주세요.
주차장은 한 시간까지 무료이고, 그 뒤로는 30분에 500원입니다.
전기차 충전기는 서문 주차장에 네 대 있습니다.

시설 이용

야외 취사와 음주는 할 수 없습니다.
드론은 날릴 수 없습니다.
분수 광장은 매시 정각에 20분 동안 가동합니다.
화장실은 정문, 동문, 호수 쉼터, 어린이 놀이터 네 곳에 있습니다.
산책로 조명은 해가 진 뒤 두 시간 동안 켜 둡니다.

문의

공원 관리 사무소는 정문 옆 1층에 있습니다.
평일 오전 9시부터 오후 7시까지 문을 엽니다.
전화는 관리 사무소 대표 번호로 걸어 주세요.
`;

export const SAMPLES = {
  code: {
    before: CODE_BEFORE,
    after: CODE_AFTER,
    beforeLabel: 'cart.js @ main',
    afterLabel: 'cart.js @ currency',
    language: 'javascript'
  },
  prose: {
    before: PROSE_BEFORE,
    after: PROSE_AFTER,
    beforeLabel: 'Draft',
    afterLabel: 'Published'
  },
  config: {
    before: CONFIG_BEFORE,
    after: CONFIG_AFTER,
    beforeLabel: 'deploy.yml @ v1',
    afterLabel: 'deploy.yml @ v2',
    language: 'yaml'
  },
  korean: {
    before: KOREAN_BEFORE,
    after: KOREAN_AFTER,
    beforeLabel: '지난해',
    afterLabel: '올해'
  },
  whitespace: {
    before: 'const total = subtotal + tax;   \n  const rounded = round(total);\nreturn rounded;\n',
    after: 'const total = subtotal + tax;\n    const rounded = round(total);\nreturn rounded;\n',
    beforeLabel: 'Before',
    afterLabel: 'After',
    language: 'javascript'
  }
} satisfies Record<string, Sample>;

export type SampleName = keyof typeof SAMPLES;

File: packages/flutter/example/lib/samples.dart
/// The pairs the gallery compares.
///
/// Written here rather than in the screens, because every demo showing the same
/// pair is what makes the options comparable: a reader turning wrapping on
/// should see the wrapping change and nothing else.
///
/// Each pair is longer than it has to be to show one change, and deliberately:
/// a demo is somewhere to scroll a comparison, fold it, search it and step
/// between changes, and a pair that fits on one screen answers none of those.
/// So the changes are spread from the top of each document to the bottom rather
/// than gathered where they would all be visible at once.
library;

import 'dart:math' as math;
import 'dart:typed_data';

import 'package:diffine/diffine.dart';

/// One pair of documents, and what to call each side.
class Sample {
  /// One pair.
  const Sample({
    required this.before,
    required this.after,
    required this.beforeLabel,
    required this.afterLabel,
    this.language = 'plain',
  });

  /// The left document.
  final String before;

  /// The right one.
  final String after;

  /// What the left one is called.
  final String beforeLabel;

  /// What the right one is called.
  final String afterLabel;

  /// What the pair is written in, for `language`.
  final String language;
}

const String _codeBefore = r'''
const double taxRate = 0.08;

class Line {
  Line({required this.sku, required this.name, required this.price, this.quantity = 1});

  final String sku;
  final String name;
  final num price;
  int quantity;
}

num subtotal(List<Line> lines) {
  num total = 0;

  for (final Line line in lines) {
    total += line.price * line.quantity;
  }

  return total;
}

num tax(List<Line> lines) {
  return subtotal(lines) * taxRate;
}

num total(List<Line> lines) {
  return subtotal(lines) + tax(lines);
}

// Rounding is the caller's problem for now.
String format(num amount) {
  return '$' + amount.toStringAsFixed(2);
}

Cart addItem(Cart cart, Product product, {int quantity = 1}) {
  final Line? found = cart.lines.where((Line line) => line.sku == product.sku).firstOrNull;

  if (found != null) {
    found.quantity += quantity;

    return cart;
  }

  cart.lines.add(
    Line(sku: product.sku, name: product.name, price: product.price, quantity: quantity),
  );

  return cart;
}

Cart removeItem(Cart cart, String sku) {
  cart.lines.removeWhere((Line line) => line.sku == sku);

  return cart;
}

Cart setQuantity(Cart cart, String sku, int quantity) {
  final Line? found = cart.lines.where((Line line) => line.sku == sku).firstOrNull;

  if (found == null) {
    return cart;
  }

  if (quantity < 1) {
    return removeItem(cart, sku);
  }

  found.quantity = quantity;

  return cart;
}

int itemCount(Cart cart) {
  return cart.lines.fold(0, (int sum, Line line) => sum + line.quantity);
}

bool isEmpty(Cart cart) {
  return cart.lines.isEmpty;
}

Summary summarise(Cart cart) {
  return Summary(
    lines: cart.lines.length,
    items: itemCount(cart),
    total: format(total(cart.lines)),
  );
}
''';

const String _codeAfter = r'''
const Map<String, int> decimals = <String, int>{'USD': 2, 'EUR': 2, 'JPY': 0, 'KRW': 0};
const Map<String, double> taxRates = <String, double>{'US': 0.08, 'DE': 0.19, 'KR': 0.1};

class Line {
  Line({
    required this.sku,
    required this.name,
    required this.price,
    required this.currency,
    this.quantity = 1,
  });

  final String sku;
  final String name;
  final num price;
  final String currency;
  int quantity;
}

num subtotal(List<Line> lines, {String currency = 'USD'}) {
  num total = 0;

  for (final Line line in lines) {
    total += line.price * line.quantity;
  }

  return round(total, currency);
}

num round(num amount, String currency) {
  return num.parse(amount.toStringAsFixed(decimals[currency] ?? 2));
}

num tax(List<Line> lines, {String region = 'US', String currency = 'USD'}) {
  return round(subtotal(lines, currency: currency) * (taxRates[region] ?? 0), currency);
}

num total(List<Line> lines, {String region = 'US', String currency = 'USD'}) {
  return round(
    subtotal(lines, currency: currency) + tax(lines, region: region, currency: currency),
    currency,
  );
}

String format(num amount, {String currency = 'USD'}) {
  return NumberFormat.simpleCurrency(name: currency).format(amount);
}

Cart addItem(Cart cart, Product product, {int quantity = 1}) {
  final Line? found = cart.lines.where((Line line) => line.sku == product.sku).firstOrNull;

  if (found != null) {
    found.quantity += quantity;

    return cart;
  }

  cart.lines.add(
    Line(
      sku: product.sku,
      name: product.name,
      price: product.price,
      currency: product.currency ?? cart.currency,
      quantity: quantity,
    ),
  );

  return cart;
}

Cart removeItem(Cart cart, String sku) {
  cart.lines.removeWhere((Line line) => line.sku == sku);

  return cart;
}

Cart setQuantity(Cart cart, String sku, int quantity) {
  final Line? found = cart.lines.where((Line line) => line.sku == sku).firstOrNull;

  if (found == null) {
    return cart;
  }

  if (quantity < 1) {
    return removeItem(cart, sku);
  }

  found.quantity = quantity;

  return cart;
}

Cart applyDiscount(Cart cart, String code) {
  final Discount? rule = cart.discounts[code];

  if (rule == null) {
    return cart;
  }

  cart.discount = Discount(code: code, off: rule.off);

  return cart;
}

int itemCount(Cart cart) {
  return cart.lines.fold(0, (int sum, Line line) => sum + line.quantity);
}

Summary summarise(Cart cart) {
  return Summary(
    lines: cart.lines.length,
    items: itemCount(cart),
    currency: cart.currency,
    total: format(
      total(cart.lines, region: cart.region, currency: cart.currency),
      currency: cart.currency,
    ),
  );
}
''';

const String _proseBefore = '''
The comparison is worked out in two passes.

Lines are matched first, which says where the changes are. Every pair of lines
inside a change is then compared again, one level down, which says what happened
inside them.

The result is one flat list of rows.

Matching the lines

The two documents are cut into lines and walked against each other. A line that
appears on both sides in the same order is left alone.

Everything else is a change: a run of lines that arrived, a run that went away,
or a run that was replaced.

Looking inside a change

A replaced run is compared again, word by word. A word is a run of letters and
digits, and everything between two words is its own token.

Where a line changed by a digit or two, the words are cut again into graphemes,
so that a number that moved is a number that moved rather than a new one.

What comes back

Each row holds a line, a kind and a number.

A viewer draws the rows in order and needs nothing else.
''';

const String _proseAfter = '''
The comparison is worked out in two passes.

Lines are matched first, which says where the changes are. Every pair of lines
inside a change is then compared a second time, one level down, which says what
happened inside them — words, or graphemes where a number moved.

The result is one flat list of rows, in document order.
Each row holds whichever side has a line on it.

Matching the lines

The two documents are cut into lines and walked against each other, longest run
of agreement first. A line that appears on both sides in the same order is left
alone.

Everything else is a change: a run of lines that arrived, a run that went away,
or a run that was replaced. A run that arrived beside one that went away is read
as a replacement rather than as two changes.

Looking inside a change

A replaced run is compared again, word by word. A word is a run of letters and
digits, and everything between two words is its own token, so a comma that moved
is a comma that moved.

Where a line changed by a digit or two, the words are cut again into graphemes,
so that a number that moved is a number that moved rather than a new one.

What comes back

Each row holds a line, a kind, a number on each side, and the pieces inside it.

A viewer draws the rows in order and needs nothing else. Nothing in the answer
is a widget, so a build script can ask for one and hand it on.
''';

/// Two versions of a source file.
const Sample code = Sample(
  before: _codeBefore,
  after: _codeAfter,
  beforeLabel: 'checkout.dart',
  afterLabel: 'checkout.dart (draft)',
  language: 'dart',
);

/// Two versions of a paragraph.
const Sample prose = Sample(
  before: _proseBefore,
  after: _proseAfter,
  beforeLabel: 'v1',
  afterLabel: 'v2',
);

/// A long file, for the folding and the virtualised rows.
Sample get longFile {
  final List<String> lines = List<String>.generate(
    400,
    (int index) => 'const step$index = ${index * 7 % 97};',
  );
  final List<String> after = List<String>.from(lines);

  after[40] = 'const step40 = 41; // adjusted';
  after[180] = 'const step180 = 3; // adjusted';
  after.insert(300, 'const extra = true;');
  after.removeAt(360);

  return Sample(
    before: '${lines.join('\n')}\n',
    after: '${after.join('\n')}\n',
    beforeLabel: 'constants.dart',
    afterLabel: 'constants.dart (draft)',
    language: 'dart',
  );
}

/// A patch, for the viewer that is handed a comparison rather than two
/// documents.
const String patch = '''
--- a/lib/src/engine.dart
+++ b/lib/src/engine.dart
@@ -12,7 +12,7 @@ class Engine {
   final List<Rule> rules;

   /// Runs every rule over the document.
-  List<Token> run(String document) {
+  List<Token> run(String document, {bool strict = false}) {
     final List<Token> out = <Token>[];

     return out;
@@ -48,6 +48,8 @@ class Engine {
   void reset() {
     _cache.clear();
   }
+
+  bool get isEmpty => _cache.isEmpty;
 }
''';

/// A drawing, for the picture comparison.
///
/// Built rather than shipped, so the gallery carries no asset and every
/// platform draws exactly the same pixels — which is also what makes the
/// comparison worth looking at: the second picture is the first one with a few
/// deliberate differences in it.
///
/// [pass] 0 is the original. 1 and 2 are two rounds of editing done to it, and
/// neither touches what the other did — which is what a comparison of three
/// pictures has to be able to show: a pane apiece, each marked where its own
/// picture left the original.
DiffPixels drawing({int pass = 0}) {
  const int width = 320;
  const int height = 200;
  final Uint8List data = Uint8List(width * height * 4);

  void paint(int x, int y, int red, int green, int blue) {
    if (x < 0 || y < 0 || x >= width || y >= height) {
      return;
    }

    final int at = (y * width + x) * 4;

    data[at] = red;
    data[at + 1] = green;
    data[at + 2] = blue;
    data[at + 3] = 255;
  }

  void box(int left, int top, int right, int bottom, int red, int green, int blue) {
    for (int y = top; y < bottom; y += 1) {
      for (int x = left; x < right; x += 1) {
        paint(x, y, red, green, blue);
      }
    }
  }

  // The ground, and a band across the top. The second pass restyled the band.
  box(0, 0, width, height, 246, 248, 251);
  box(0, 0, width, 36, pass == 2 ? 18 : 27, pass == 2 ? 58 : 34, pass == 2 ? 92 : 44);

  // Three cards.
  for (int card = 0; card < 3; card += 1) {
    final int left = 16 + card * 98;

    box(left, 56, left + 82, 130, 255, 255, 255);
    box(left, 56, left + 82, 60, 14, 127, 252);

    // A line of "text" in each. The first pass lengthened the middle card's,
    // the second cut the first card's back.
    for (int line = 0; line < 3; line += 1) {
      final int width =
          60 - line * 14 + (pass == 1 && card == 1 ? 10 : 0) - (pass == 2 && card == 0 ? 16 : 0);

      box(left + 10, 74 + line * 14, left + 10 + width, 80 + line * 14, 200, 208, 220);
    }
  }

  // A circle at the bottom, moved a little by the first pass.
  final int centre = pass == 1 ? 178 : 160;

  for (int y = 148; y < 188; y += 1) {
    for (int x = centre - 20; x < centre + 20; x += 1) {
      final double distance = math.sqrt(math.pow(x - centre, 2) + math.pow(y - 168, 2));

      if (distance <= 18) {
        paint(x, y, pass == 1 ? 232 : 26, pass == 1 ? 62 : 127, pass == 1 ? 140 : 75);
      }
    }
  }

  return DiffPixels(data: data, width: width, height: height);
}
